# HG changeset patch
# User Rik <rik@octave.org>
# Date 1519688583 28800
#      Mon Feb 26 15:43:03 2018 -0800
# Branch stable
# Node ID d3a79cb734d2ca32f2a4ca22fc4d368037cbbca0
# Parent  956e28867c80e826ecfbac04c6091ad71e96edf2
Fix performance of Sparse fsolve for complex sparse matrices (bug #53140).

* CSparse.cc (SparseComplexMatrix::bsolve, SparseComplexMatrix::fsolve): Use
floating point arithmetic to calculate a new matrix size as a fraction of the
existing number of non-zero elements when a resize is required.

diff --git a/liboctave/array/CSparse.cc b/liboctave/array/CSparse.cc
--- a/liboctave/array/CSparse.cc
+++ b/liboctave/array/CSparse.cc
@@ -4622,19 +4622,20 @@ SparseComplexMatrix::bsolve (MatrixType 
                       for (octave_idx_type i = 0; i < b_nr; i++)
                         {
                           Complex tmp = Bx[i];
                           if (tmp != 0.0)
                             {
                               if (ii == x_nz)
                                 {
                                   // Resize the sparse matrix
-                                  octave_idx_type sz = x_nz *
-                                                       (b_nc - j) / b_nc;
-                                  sz = (sz > 10 ? sz : 10) + x_nz;
+                                  octave_idx_type sz;
+                                  sz = (static_cast<double> (b_nc) - j) / b_nc
+                                       * x_nz;
+                                  sz = x_nz + (sz > 100 ? sz : 100);
                                   retval.change_capacity (sz);
                                   x_nz = sz;
                                 }
                               retval.xdata (ii) = tmp;
                               retval.xridx (ii++) = i;
                             }
                         }
                       retval.xcidx (j+1) = ii;
@@ -5994,18 +5995,20 @@ SparseComplexMatrix::fsolve (MatrixType 
                   for (octave_idx_type i = 0; i < b_nr; i++)
                     {
                       Complex tmp = Xx[i];
                       if (tmp != 0.0)
                         {
                           if (ii == x_nz)
                             {
                               // Resize the sparse matrix
-                              octave_idx_type sz = x_nz * (b_nc - j) / b_nc;
-                              sz = (sz > 10 ? sz : 10) + x_nz;
+                              octave_idx_type sz;
+                              sz = (static_cast<double> (b_nc) - j) / b_nc
+                                   * x_nz;
+                              sz = x_nz + (sz > 100 ? sz : 100);
                               retval.change_capacity (sz);
                               x_nz = sz;
                             }
                           retval.xdata (ii) = tmp;
                           retval.xridx (ii++) = i;
                         }
                     }
                   retval.xcidx (j+1) = ii;
@@ -6495,18 +6498,20 @@ SparseComplexMatrix::fsolve (MatrixType 
                   for (octave_idx_type i = 0; i < b_nr; i++)
                     {
                       Complex tmp = Xx[i];
                       if (tmp != 0.0)
                         {
                           if (ii == x_nz)
                             {
                               // Resize the sparse matrix
-                              octave_idx_type sz = x_nz * (b_nc - j) / b_nc;
-                              sz = (sz > 10 ? sz : 10) + x_nz;
+                              octave_idx_type sz;
+                              sz = (static_cast<double> (b_nc) - j) / b_nc
+                                   * x_nz;
+                              sz = x_nz + (sz > 100 ? sz : 100);
                               retval.change_capacity (sz);
                               x_nz = sz;
                             }
                           retval.xdata (ii) = tmp;
                           retval.xridx (ii++) = i;
                         }
                     }
                   retval.xcidx (j+1) = ii;
